README

Uebungen
========

1. Es sei das folgende kleine Programm in C gegeben:

   int max(int i, int j)
   /* liefert das Maximum der Zahlen i und j */
   {
     return i>j?i:j;
   }

   Geben Sie die Token an, die ein Scanner fuer C liefern wuerde.
   Folgende Elemente sind Token: int, max, (, ), i, j, { ,} , return, >, ?, : ; , /*, */
	In höheren Programmiersprachen muss die Lexikalische Analyse
	typischerweise folgende Tokens erkennen:
	- Bezeichner (Identifier)
	- Operator
	- Literal
	- Kommentar
	- Schlüsselwort

	Liste von Tokens:
		SchlüEselwort(int), Bezeichner(max), Operator('('),
		SchlüEselwort(int), Bezeichner(i), Operator(','),
		SchlüEselwort(int), Bezeichner(j), Operator(')'),
		Kommentar(/* liefert das Maximum der Zahlen i und j */),
		Operator('{'),
		SchlüEselwort(return),
		Bezeichner(i), Operator('>'), Bezeichner(j), Operator('?'),
		Bezeichner(i), Operator(':'), Bezeichner(j), Operator(';'),
		Operator('}').

2. Benutzen Sie Thompson's Konstruktion, um fuer die folgenden
   regulaeren Ausdruecke NEAs zu erstellen:
   a) (a|b)*
   b) (a*|b*)*
   c) (a?b*)*
   d) (a|b)*abb(a|b)*
   Geben Sie an, welche Bewegungen jeder Automat macht, wenn der
   Eingabestring ababbab abgearbeitet wird.

3. Konvertieren Sie die NEAs aus Uebung 2 in DEAs. Geben Sie an,
   welche Bewegungen jeder Automat macht, wenn der Eingabestring
   aababbab abgearbeitet wird.


Programmieraufgabe
==================

Problemstellung
---------------

Konstruktion des Scanners fuer den SPL-Compiler.


Vorgegebene Dateien
-------------------

README          -  die Beschreibung, die Sie gerade lesen
Makefile        -  ein Makefile zur automatischen Uebersetzung
Tests           -  ein Subdirectory, in dem Sie Ihre Tests aufbewahren
common.h        -  haeufig gebrauchte Definitionen
main.c          -  das Hauptprogramm des entstehenden Compilers
preliminary.h   -  die vorlaeufige Definition der Tokens und ihrer Werte
scanner.h       -  Interface zum Scanner
scanner.l       -  die Scanner-Spezifikation: das ist Ihre Aufgabe
utils.c         -  haeufig gebrauchte Hilfsfunktionen
utils.h         -  Interface zu utils.c


Aufgaben
--------

1. Studieren Sie die SPL-Sprachdefinition.
   Machen Sie sich mit unserer Sprache vertraut, indem Sie mindestens
   5 kleine SPL-Programme schreiben. Versuchen Sie, moeglichst viele
   Features der Sprache zu benutzen. Hier ein Vorschlag (es wird
   nachdruecklich empfohlen, diesen Vorschlag aufzugreifen, da die
   verschiedenen Teilaufgaben jeweils andere Spracheigenschaften
   nutzen, so dass Sie die entstehenden Programme spaeter auch als
   Testprogramme fuer Ihren Compiler einsetzen koennen):
   a) Im Hauptprogramm: Eingabe von zwei Zahlen, Ausgabe der Summe.
   b) Dasselbe, aber mit Berechnung der Summe in einer Prozedur.
   c) Einlesen eines Arrays von Zahlen im Hauptprogramm, Sortieren
      des Arrays in einer Prozedur, Ausgabe des sortierten Arrays
      im Hauptprogramm.
   d) Berechnung von n! in einer Prozedur, sowohl iterativ als auch
      rekursiv.
   e) Zeichnen einer kleinen Figur auf dem Grafikbildschirm.
   Hinweis: Um Ihre Programme auf syntaktische Richtigkeit testen zu
   koennen, stehen Ihnen auf dem Sun-Rechner 'saturn' unter dem Pfad
   /home/hg53/spl Referenzimplementierungen fuer Linux und SunOS zur
   Verfuegung.

2. Schreiben Sie die Spezifikation eines Scanners fuer SPL.
   Ergaenzen Sie dazu die Datei 'scanner.l' und fuegen Sie eine
   Moeglichkeit zur textuellen Ausgabe von Tokens (ggf. mit Wert)
   hinzu.

3. Lassen Sie Ihren Scanner die 5 Testprogramme aus dem ersten
   Aufgabenteil analysieren. Beobachten Sie den Tokenstrom und
   korrigieren Sie ggf. Ihre Scanner-Spezifikation.

4. Schreiben Sie eine Eingabedatei zum systematischen Test ihres
   Scanners, vor allem in Hinblick auf die verschiedenen Formen
   von Ganzzahl-Literalen.
------------------------------------------------------------------------------

* 1. Studie der SPL-Sprachdefinition.

int 				Primitiver Typ für ganze Zahlen
int[] 				Typkonstruktor für Felder
proc 				Für Prozeduren
proc wert(i: int)		Prozedur mit Wertparametern
proc ref(ref j: int)		Prozedur mit Referenzparametern
if( ) 				Einarmige Anweisung
if ( )  else 			Zweiarmige Anweisung
while (i < 15)			Abweisende Schleife
type a = array [5] of int;	Typvereinbarung. type <name> = <typ> ;
array [3] of array [5] of int	Zuweisung eines mehrdimensionalen Zahlenfeldes
wert();				Einfache Prozeduraufrufe
proc(i: int){ return proc(5); }	Nicht Erlaubt -- Keine Rekursionen
proc() Proc()			Gross- und Kleinschreibung wird unterschieden
LEER,TAB,\n			Zur Token-trennung
// 				Das ist ein Kommentar
das_ist_ein_Bezeichner		Bezeichner fangen mit [a-Z_] an.
1234  0x1a2f3F4e 'a' '\n'	Werte = Dezimale, Hexadezimale und ASCII-Codes
array else if of proc		Reservierte Schlüsselwörter
ref type var while		Reservierte Schlüsselwörter
(  )  [  ]  {  }  =  #		Sprachelemente von Bedeutung
<  <=  >  >=  :=  :  ,		Sprachelemente von Bedeutung
;  +  -  *  /			Sprachelemente von Bedeutung
main()				Jedes Programm enthält mind. eine main()

------------------------------------------------------------------------------

# Token-Bezeichner:

Beispiel fuer gleiche Typen:
	type typ1 = array [5] of int;
	type typ2 = typ1;
(diese können vor einer Prozedur deklariert werden)

Prozedurvereinbarungen:
proc <name> ( <parameterliste> ) { <deklarationen> <anweisungsliste> }
proc <name> ( <name> : <typ>, ref <name> : <typ> )
{
	<deklarationen> <anweisungsliste>
}

Die Zuweisung hat die Form:
	<lhs> := <rhs> ;
	x[3] := x[2] * 2;
Wobei lhs und rhs den gleichen typ haben müssen.

Die bedingte Anweisung kann in einer von zwei Formen vorliegen:
  if ( <ausdruck> ) <anweisung1>
  if ( <ausdruck> ) <anweisung1> else <anweisung2>


fdgr46/ChangeLog
2011-12-02 03:55:37 +0100 Fernandos 

	* who knows if the tests target for the makefile is neccessary or not, re-adding it (HEAD, origin/fdgr46, fdgr46)

2011-12-02 03:52:11 +0100 Fernandos 

	* fix when an intial 'make verify' didn't count successfull programs

2011-12-02 00:08:08 +0100 Fernandos 

	* readd old testfiles (v4.3)

2011-12-02 00:07:43 +0100 Fernandos 

	* renaming test files correctly * fixing little bug in verify script and parser rule faktor_expr

2011-11-29 23:29:18 +0100 Fernandos 

	* fixed spurious progTree assignment * scanner.l now recognizes "#" * [intExp bug] scanner.l can convert ascii to integer correctly now. '\n' & 'c' is supported by now * Makefile supports these targets now: ast, run, verify, fast, clean, dist-clean * parser.y replaced bool_expr with expression and expression with arith_expression * parsery. useless emptyStmt() calls in AST, caused by missing SEMIC * removing temporary .absyn files (v4.2)

2011-11-29 23:16:41 +0100 Fernandos 

	* ignore fdgr46.tar.gz

2011-11-29 23:15:50 +0100 Fernandos 

	* replacing ignores for make verify output files * adding .absyn files to ignores

2011-11-29 23:12:57 +0100 Fernandos 

	* start ignoring

2011-11-29 15:39:58 +0100 Fernandos 

	* fertig

2011-11-29 14:50:22 +0100 Fernandos 

	* verify script enhanced almost finished. one bug with uminusexp detected and if()

2011-11-24 16:22:27 +0100 Fernandos 

	* updated gitignore to reflect branch changes

2011-11-24 16:19:49 +0100 Fernandos 

	* Merge commit 'd2135f60d97d51fb59caa09e076d8e1cce6bf110' into fdgr46

2011-11-24 16:05:02 +0100 Fernandos 

	* almost our release version, missing team.txt

2011-10-13 23:08:04 +0200 Fernandos 

	* added ChangeLog to gitignore

2011-10-13 23:08:04 +0200 Fernandos 

	* initial version of homework-1

2011-10-12 18:49:11 +0200 Fernandos 

	* added .gitignore rules (origin/master, origin/HEAD, master)

2011-10-12 16:59:15 +0200 Fernandos 

	* first commit



absyn.h
/*
 * absyn.h -- abstract syntax
 */


#ifndef _ABSYN_H_
#define _ABSYN_H_


#define ABSYN_NAMETY		0
#define ABSYN_ARRAYTY		1
#define ABSYN_TYPEDEC		2
#define ABSYN_PROCDEC		3
#define ABSYN_PARDEC		4
#define ABSYN_VARDEC		5
#define ABSYN_EMPTYSTM		6
#define ABSYN_COMPSTM		7
#define ABSYN_ASSIGNSTM		8
#define ABSYN_IFSTM		9
#define ABSYN_WHILESTM		10
#define ABSYN_CALLSTM		11
#define ABSYN_OPEXP		12
#define ABSYN_VAREXP		13
#define ABSYN_INTEXP		14
#define ABSYN_SIMPLEVAR		15
#define ABSYN_ARRAYVAR		16
#define ABSYN_DECLIST		17
#define ABSYN_STMLIST		18
#define ABSYN_EXPLIST		19

#define ABSYN_OP_EQU		0
#define ABSYN_OP_NEQ		1
#define ABSYN_OP_LST		2
#define ABSYN_OP_LSE		3
#define ABSYN_OP_GRT		4
#define ABSYN_OP_GRE		5
#define ABSYN_OP_ADD		6
#define ABSYN_OP_SUB		7
#define ABSYN_OP_MUL		8
#define ABSYN_OP_DIV		9


typedef struct absyn {
  int type;
  int line;
  union {
    struct {
      Sym *name;
    } nameTy;
    struct {
      int size;
      struct absyn *ty;
    } arrayTy;
    struct {
      Sym *name;
      struct absyn *ty;
    } typeDec;
    struct {
      Sym *name;
      struct absyn *params;
      struct absyn *decls;
      struct absyn *body;
    } procDec;
    struct {
      Sym *name;
      struct absyn *ty;
      boolean isRef;
    } parDec;
    struct {
      Sym *name;
      struct absyn *ty;
    } varDec;
    struct {
      int dummy;		/* empty struct not allowed in C */
    } emptyStm;
    struct {
      struct absyn *stms;
    } compStm;
    struct {
      struct absyn *var;
      struct absyn *exp;
    } assignStm;
    struct {
      struct absyn *test;
      struct absyn *thenPart;
      struct absyn *elsePart;
    } ifStm;
    struct {
      struct absyn *test;
      struct absyn *body;
    } whileStm;
    struct {
      Sym *name;
      struct absyn *args;
    } callStm;
    struct {
      int op;
      struct absyn *left;
      struct absyn *right;
    } opExp;
    struct {
      struct absyn *var;
    } varExp;
    struct {
      int val;
    } intExp;
    struct {
      Sym *name;
    } simpleVar;
    struct {
      struct absyn *var;
      struct absyn *index;
    } arrayVar;
    struct {
      boolean isEmpty;
      struct absyn *head;
      struct absyn *tail;
    } decList;
    struct {
      boolean isEmpty;
      struct absyn *head;
      struct absyn *tail;
    } stmList;
    struct {
      boolean isEmpty;
      struct absyn *head;
      struct absyn *tail;
    } expList;
  } u;
} Absyn;

/* Type constructors */
Absyn *newNameTy(int line, Sym *name);
Absyn *newArrayTy(int line, int size, Absyn *ty);

/* Type declaration constructors */
Absyn *newTypeDec(int line, Sym *name, Absyn *ty);
Absyn *newProcDec(int line, Sym *name,
                  Absyn *params, Absyn *decls, Absyn *body);

/* Variable declaration constructors */
Absyn *newParDec(int line, Sym *name, Absyn *ty, boolean isRef);
Absyn *newVarDec(int line, Sym *name, Absyn *ty);

/* Statement constructors */
Absyn *newEmptyStm(int line);
Absyn *newCompStm(int line, Absyn *stms);
Absyn *newAssignStm(int line, Absyn *var, Absyn *exp);
Absyn *newIfStm(int line, Absyn *test, Absyn *thenPart, Absyn *elsePart);
Absyn *newWhileStm(int line, Absyn *test, Absyn *body);
Absyn *newCallStm(int line, Sym *name, Absyn *args);

/* Expression constructors */
Absyn *newOpExp(int line, int op, Absyn *left, Absyn *right);
Absyn *newVarExp(int line, Absyn *var);
Absyn *newIntExp(int line, int val);

/* Variable constructors */
Absyn *newSimpleVar(int line, Sym *name);
Absyn *newArrayVar(int line, Absyn *var, Absyn *index);

/* List constructors */
Absyn *emptyDecList(void);
Absyn *newDecList(Absyn *head, Absyn *tail);
Absyn *emptyStmList(void);
Absyn *newStmList(Absyn *head, Absyn *tail);
Absyn *emptyExpList(void);
Absyn *newExpList(Absyn *head, Absyn *tail);

void showAbsyn(Absyn *node);


#endif /* _ABSYN_H_ */
